app/src/main/java/com/example/a7asryan/register/viewmodel/RegisterViewModel.kt

```diff
@@ -9,20 +11,89 @@ import kotlinx.coroutines.launch
 
 class RegisterViewModel(private val repository: Repository) : ViewModel() {
 
-    private var _isExistence:MutableLiveData<Boolean> = MutableLiveData()
-    val isExistence = _isExistence
+    private val _registerResult: MutableLiveData<RegisterResult> = MutableLiveData()
+    val registerResult: LiveData<RegisterResult>
+        get() = _registerResult
 
-    fun checkUserExistence(email: String, password: String) {
+    fun registerUser(email: String, password: String, userName: String, phoneNumber: String) {
+        when {
+            isEmailInvalid(email) -> {
+                _registerResult.value = RegisterResult.InvalidData(ErrorType.EmailError)
+            }
+            isPasswordInvalid(password) -> {
+                _registerResult.value = RegisterResult.InvalidData(ErrorType.PasswordError)
+            }
+            isUserNameInvalid(userName) -> {
+                _registerResult.value = RegisterResult.InvalidData(ErrorType.UserNameError)
+            }
+            isPhoneNumberInvalid(phoneNumber) -> {
+                _registerResult.value = RegisterResult.InvalidData(ErrorType.PhoneNumberError)
+            }
+            else -> {
+                checkUserExistence(User(userName, email, password, phoneNumber))
+            }
+        }
+    }
+
+    private fun checkUserExistence(user: User) {
         viewModelScope.launch {
-            val response = repository.checkUser(email, password)
-            _isExistence.postValue(response)
+            val response = repository.checkUser(user.email, user.password)
+            if (response) {
+                _registerResult.value =RegisterResult.RegisterError
+            } else {
+                insertUser(user)
+                _registerResult.value =RegisterResult.RegisterSuccessful
+            }
         }
     }
 
-    fun insertUser(user: User) {
+    private fun insertUser(user: User) {
         viewModelScope.launch {
```

Comment:
This can be `suspend` , you are launching twice.

Reply:
You are right ^^